src/main/java/com/epam/aidial/core/storage/BlobWriteStream.java

```diff
@@ -99,33 +112,27 @@ public void end(Handler<AsyncResult<Void>> handler) {
                     throw new RuntimeException(exception);
                 }
 
-                Buffer lastChunk = chunkBuffer.slice(0, position);
-                metadata = new FileMetadata(resource, bytesHandled, contentType);
+                byte[] lastChunk = chunkBuffer.slice(0, position).getBytes();
                 if (mpu == null) {
                     log.info("Resource is too small for multipart upload, sending as a regular blob");
-                    storage.store(resource.getAbsoluteFilePath(), contentType, lastChunk);
+                    metadata = resourceService.putFile(resource, lastChunk, etag, contentType);
                 } else {
                     if (position != 0) {
                         MultipartPart part = storage.storeMultipartPart(mpu, ++chunkNumber, lastChunk);
                         parts.add(part);
                     }
 
-                    storage.completeMultipartUpload(mpu, parts);
+                    String newEtag = etagBuilder.append(lastChunk).build();
+                    metadata = (FileMetadata) new FileMetadata(resource, bytesHandled, contentType).setEtag(newEtag);
+                    mpu.blobMetadata().getUserMetadata().put(ResourceUtil.ETAG_ATTRIBUTE, newEtag);
+                    resourceService.completeMultipartUpload(resource, mpu, parts, etag);
```

Comment:
Please name two methods similarly. For example, beginFileUpload and finishFileUpload

Reply:
Done.